source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -100,6 +100,20 @@ FilterHeadersStatus Filter::decodeHeaders(RequestHeaderMap& headers, bool end_st
 
 Http::FilterDataStatus Filter::onData(ProcessorState& state, ProcessingMode::BodySendMode body_mode,
                                       Buffer::Instance& data, bool end_stream) {
+  if (state.callbackState() == ProcessorState::CallbackState::Headers) {
+    ENVOY_LOG(trace, "Header processing still in progress -- holding body data");
+    if (end_stream) {
+      // Buffer the data and we'll send the callback when the response callback returns.
+      state.setBodySendDeferred(true);
+      return FilterDataStatus::StopIterationAndBuffer;
+    } else {
+      // We're not ready to decide whether we want to buffer the body, but we
+      // also don't want to fill the buffer, so pause receiving additional chunks.
```

Comment:
Not immediately clear why we know to buffer the data in the end_stream case, can you elaborate?

Reply:
I tried to clarify these comments.